pkg/controller/login/select_password.go

```diff
@@ -81,36 +71,35 @@ func (c *Controller) HandleSubmitNewPassword() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		ctx := r.Context()
 		session := controller.SessionFromContext(ctx)
-		flash := flash.New(session.Values)
+		if session == nil {
+			controller.MissingSession(w, r, c.h)
+			return
+		}
+		flash := controller.Flash(session)
 
 		code := r.FormValue("oobCode")
 
 		var form FormData
 		if err := controller.BindForm(w, r, &form); err != nil {
 			flash.Error("Select password failed: %v", err)
-			c.renderShowSelectPassword(ctx, w, "", code, false, flash)
+			c.renderShowSelectPassword(ctx, w, "", code, false)
 			return
 		}
 		email := strings.TrimSpace(form.Email)
 
 		if err := c.validateComplexity(form.Password); err != nil {
 			flash.Error("Select password failed: %v", err)
-			c.renderShowSelectPassword(ctx, w, email, code, false, flash)
+			c.renderShowSelectPassword(ctx, w, email, code, false)
 			return
 		}
 
-		if _, err := c.firebaseInternal.ChangePasswordWithCode(ctx, code, form.Password); err != nil {
-			if errors.Is(err, firebase.ErrInvalidOOBCode) || errors.Is(err, firebase.ErrExpiredOOBCode) {
-				flash.Error("The action code is invalid. This can happen if the code is malformed, expired, or has already been used.")
-				c.renderShowSelectPassword(ctx, w, email, code, true, flash)
-			} else {
-				flash.Error("Select password failed. %v", err)
-				c.renderShowSelectPassword(ctx, w, email, code, false, flash)
-			}
+		if err := c.authProvider.ChangePassword(ctx, form.Password, code); err != nil {
+			flash.Error("Failed to change password: %v", err)
+			c.renderShowSelectPassword(ctx, w, email, code, true)
 			return
 		}
 
-		if err := c.db.PasswordChanged(email, time.Now()); err != nil {
+		if err := c.db.PasswordChanged(email, time.Now().UTC()); err != nil {
```

Comment:
Felt like a bug?

Reply:
This is better, but I think the db retains the local timezone info, so when we later subtract to get time since last pwd change,  it still works.